### PR TITLE
[BEAM-8269] Convert Py3 type hints to Beam types

### PR DESCRIPTION
**Pull request by udim**
Also adds test coverage for strip_iterable and fixes a couple of cases
in get_yielded_type.

And other fixes (see commit comments).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
Still testing internally, do not merge yet

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
@robertwb PTAL, changes:
- Commented out one test case with a TODO(BEAM-8492) comment.
- Added a _LOGGER.info message when converting an unknown typing module type to typehints.Any.

**Comment by udim** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/typehints/decorators_test_py3.py [hunk 1]
```diff
@@ -33,18 +34,20 @@
...
+# Name is 'T' so it converts to a beam type with the same name.
+T_typing = typing.TypeVar('T')
 
 
 class IOTypeHintsTest(unittest.TestCase):
 
   def test_from_callable(self):
     def fn(a: int, b: str = None, *args: Tuple[T], foo: List[int],
-           **kwargs: Dict[str, str]) -> Tuple:
+           **kwargs: Dict[str, str]) -> Tuple[Any, ...]:
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/decorators_test_py3.py]: These should be synonyms, right?

File: sdks/python/apache_beam/typehints/typehints.py [hunk 2]
```diff
@@ -1190,9 +1190,14 @@ def get_yielded_type(type_hint):
...
     return type_hint.yielded_type
   if is_consistent_with(type_hint, Tuple[Any, ...]):
-    return Union[type_hint.tuple_types]
+    if isinstance(type_hint, TupleConstraint):
+      return Union[type_hint.tuple_types]
+    else:  # TupleSequenceConstraint
+      return type_hint.inner_type
   if is_consistent_with(type_hint, Iterable[Any]):
     return type_hint.inner_type
+  if is_consistent_with(type_hint, Dict[Any, Any]):
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typehints.py]: Supporting this feels like a bug. Do we guard against it elsewhere?

**Comment by robertwb** [comment]
BTW, I'm not able to reproduce the error, in 2.7 or 3.6, with or without cython.

**Comment by udim** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/typehints/decorators_test_py3.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/decorators_test_py3.py]: Yes if these were typing module types, but in this case this is `typehints.Tuple[typehints.Any, ...]`, which requires subscripting to turn it into a constraint. I will remove such usage in the future and make the `typehints.` prefix explicit.

FYI, the conversion from typing.Tuple to a typehints.Tuple constraint is taken care of here: https://github.com/apache/beam/pull/10042/files#diff-9b2eac354738047adddd44814d4b429cR245

File: sdks/python/apache_beam/typehints/typehints.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typehints.py]: You're right, this branch has been removed.

**Comment by udim** [comment]
Fixed `test_strip_iterable` failure in latest force push.

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Pull request merged**